client/lib/html-toolbar/index.js

```diff
@@ -0,0 +1,229 @@
+ * @returns {Object} The split textarea content.
+ */
+export const splitSelectedContent = textarea => {
+	const { selectionEnd, selectionStart, value } = textarea;
+	return {
+		before: value.substring( 0, selectionStart ),
+		inner: value.substring( selectionStart, selectionEnd ),
+		after: value.substring( selectionEnd, value.length ),
+	};
+};
+
+/**
+ * Set the cursor position after the insertion of new content into a textarea.
+ *
+ * @param {Element} textarea The editor textarea.
+ * @param {Number} selectionEnd The textarea.selectionEnd value BEFORE the content insertion.
+ * @param {Number} insertedContentLength The length of the inserted content.
+ */
+export const setCursorPosition = ( textarea, selectionEnd, insertedContentLength ) => {
+	textarea.selectionEnd = textarea.selectionStart = selectionEnd + insertedContentLength;
+};
+
+/**
+ * Insert content into a textarea.
+ *
+ * @param {Element} textarea The target textarea.
+ * @param {String} before The content before the selection.
+ * @param {String} content The new content to insert.
+ * @param {String} after The content after the selection.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertContent = ( textarea, before, content, after, onInsert = noop ) => {
+	const { selectionEnd, value } = textarea;
+	const newContent = before + content + after;
+	const userAgent = get( window, 'navigator.userAgent', '' );
+
+	if ( /(?:firefox|fxios)/i.test( userAgent ) ) {
+		// In Firefox, execCommand( 'insertText' ), needed to preserve the undo stack,
+		// always moves the cursor to the end of the content.
+		// A workaround involving manually setting the cursor position and inserting the editor content
+		// is needed to put the cursor back to the correct position.
+		textarea.value = newContent;
+		textarea.focus();
+		document.execCommand( 'insertText', false, content );
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		textarea.focus();
+		onInsert( newContent );
+		return;
+	} else if ( /(?:edge|msie |trident.+?; rv:)/i.test( userAgent ) ) {
+		// execCommand( 'insertText' ), needed to preserve the undo stack, does not exist in IE11,
+		// and Edge has issues positioning the cursor after insertion.
+		textarea.value = newContent;
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		onInsert( newContent );
+		return;
+	}
+	textarea.focus();
+	document.execCommand( 'insertText', false, content );
+};
+
+/**
+ * Convert an object of attributes into a string.
+ *
+ * @param {Object} attributes The object of attributes.
+ * @returns {String} The string of attributes.
+ */
+export const attributesToString = ( attributes = {} ) =>
+	reduce(
+		attributes,
+		( attributesString, attributeValue, attributeName ) =>
+			attributeValue
+				? attributesString + ` ${ attributeName }="${ attributeValue }"`
+				: attributesString,
+		''
+	);
```

Comment:
This algorithm suffers from the dangling-delimiter-dilemna - does the extra space go at the front or at the back?

```js
export const attributesToString = ( attributes = {} ) =>
	map( attributes, ( value, name ) => `${ name }="${ value }"` ).join( ' ' )
```